src/queues/mod.rs

```diff
@@ -0,0 +1,212 @@
+}
+
+type SendFuture = Box<Future<Item = String, Error = QueueError>>;
+
+pub trait Factory {
+    fn new(id: String, settings: &Settings) -> Self;
+}
+
+#[derive(Debug, Default)]
+pub struct Message {
+    pub id: String,
+    pub notification: Notification,
+}
+
+#[derive(Debug)]
+pub struct QueueError {
+    description: String,
+}
+
+#[derive(Debug)]
+pub struct QueueIds {
+    pub bounce: String,
+    pub complaint: String,
+    pub delivery: String,
+    pub notification: String,
+}
+
+impl Queues {
+    pub fn new<Q: 'static>(ids: QueueIds, settings: &Settings) -> Queues
+    where
+        Q: Incoming + Outgoing + Factory,
+    {
+        Queues {
+            bounce: Box::new(Q::new(ids.bounce, settings)),
+            complaint: Box::new(Q::new(ids.complaint, settings)),
+            delivery: Box::new(Q::new(ids.delivery, settings)),
+            notification: Box::new(Q::new(ids.notification, settings)),
+            db: DbClient::new(settings),
+        }
+    }
+
+    pub fn process(&'static self) -> QueueFuture {
+        let joined_futures = self
+            .process_queue(&self.bounce)
+            .join3(
+                self.process_queue(&self.complaint),
+                self.process_queue(&self.delivery),
+            )
+            .map(|results| results.0 + results.1 + results.2);
+
+        Box::new(joined_futures)
+    }
+
+    fn process_queue(&'static self, queue: &'static Box<Incoming>) -> QueueFuture {
+        let future = queue
+            .receive()
+            .and_then(move |messages| {
+                let mut futures: Vec<Box<Future<Item = (), Error = QueueError>>> = Vec::new();
+                for message in messages.into_iter() {
+                    if message.notification.notification_type != NotificationType::Null {
+                        let future = self
+                            .handle_notification(&message.notification)
+                            .and_then(move |_| queue.delete(message));
+                        futures.push(Box::new(future));
+                    }
+                }
+                future::join_all(futures.into_iter())
+            })
+            .map(|results| results.len());
+        Box::new(future)
+    }
+
+    fn handle_notification(
+        &'static self,
+        notification: &Notification,
+    ) -> Box<Future<Item = (), Error = QueueError>> {
+        let result = match notification.notification_type {
+            NotificationType::Bounce => self.record_bounce(notification),
+            NotificationType::Complaint => self.record_complaint(notification),
+            NotificationType::Delivery => Ok(()),
+            NotificationType::Null => {
+                Err(QueueError::new(String::from("Insane notification type")))
+            }
+        };
+        match result {
+            Ok(_) => {
+                let future = self
+                    .notification
+                    .send(&notification)
```

Comment:
So IIUC, we're basically pulling SES notifications from three separate "incoming" queues, recording them in the db, and then forwarding them over to the auth-server via a single "outgoing" queue.  Is that accurate?

I wonder whether it's worth teasing the "incoming" and "outgoing" queues apart a little more, because the incoming queues are and will always be very specific to the details of SES, but ISTM the outgoing queue should be more generic and hide some of those details.  For example, does it make sense to have an outgoing `Notification` type that doesn't include SES-specific details like `mail.source_arn`?

Reply:
> For example, does it make sense to have an outgoing `Notification` type that doesn't include SES-specific details like `mail.source_arn`?

Yep. This is what I meant in the opening comment about the `Notification` format leaking out. It was a shortcut taken to try and reduce the volume of data-marshalling in this code review, and I was thinking YAGNI (at least for now, anyway). Happy to add it though.

Reply:
> I wonder whether it's worth teasing the "incoming" and "outgoing" queues apart a little more, because the incoming queues are and will always be very specific to the details of SES, but ISTM the outgoing queue should be more generic and hide some of those details.

Just to add a bit more about this part of your comment, because I don't think my intentions were clear. I was aiming for something that can plug in to the existing auth server code and have it just work without many changes needed over there. Because the auth server already has code for emitting flow events from SQS notifications, I figured the natural first step was to get everything working end-to-end using the same message format. As a secondary consideration it was also nice that it made this PR a bit simpler, because it's one less data mapping to implement, test and review.

I agree it's definitely not what we want to do long-term, but quite like the idea of limiting the amount of short-term churn we'll have when we first come to glue everything together.

In terms of teasing them apart some more right now, would it suffice for `Outgoing` to just take a json value? It would just be the direct serialisation of the existing `Notification` type in this PR, but at least it would then be decoupled for the future change when we do come to send our own, more-generic message format.

Reply:
Thanks, that makes sense.  I agree there's definitely value in having something that will plug in directly to the auth-server as-is in order to evolve things forward.  I wonder if we can just send a Notification but without any bits that the auth-server won't read, e.g. an "OutgoingNotification" type that has a subset of the fields of the incoming "Notification" type.